tests/check.rs

```diff
@@ -212,3 +212,71 @@ fn build_check() {
     assert_that(foo.cargo_process("check"),
                 execs().with_status(0));
 }
+
+// Checks that where a project has both a lib and a bin, the lib is only checked
+// not built.
+#[test]
+fn issue_3418() {
+    if !is_nightly() {
+        return;
+    }
+
+    let foo = project("foo")
+        .file("Cargo.toml", r#"
+            [package]
+            name = "foo"
+            version = "0.1.0"
+            authors = []
+
+            [dependencies]
+        "#)
+        .file("src/lib.rs", "")
+        .file("src/main.rs", "fn main() {}");
+    foo.build();
+
+    assert_that(foo.cargo_process("check").arg("-v"),
+                execs().with_status(0)
+                       .with_stderr_does_not_contain("--crate-type lib"));
+}
+
+// Some weirdness that seems to be caused by a crate being built as well as
+// checked, but in this case with a proc macro too.
+#[test]
+fn issue_3419() {
+    if !is_nightly() {
+        return;
+    }
+
+    let foo = project("foo")
+        .file("Cargo.toml", r#"
+            [package]
+            name = "foo"
+            version = "0.0.1"
+            authors = []
+
+            [dependencies]
+            rustc-serialize = "*"
```

Comment:
Also out of curiosity, what was the bug here? Did fixing the profiles fix it? If so that seems odd...


Unless we picked up the in-tree copy of rustc-serialize... oh dear...